### PR TITLE
Fixed what-if functionality with -QueryString parameter

### PR DESCRIPTION
**Pull request by Xynoclafe**
## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Resources/ResourceManager/SdkModels/Deployments/PSDeploymentWhatIfCmdletParameters.cs [hunk 1]
```diff
@@ -102,6 +102,11 @@ public DeploymentWhatIf ToDeploymentWhatIf()
             else if (Uri.IsWellFormedUriString(this.TemplateUri, UriKind.Absolute))
             {
                 properties.TemplateLink = new TemplateLink(this.TemplateUri);
+
+                if (!string.IsNullOrEmpty(this.QueryString))
+                {
+                    properties.TemplateLink.QueryString = this.QueryString;
```
Comment (shenglol) [inline src/Resources/ResourceManager/SdkModels/Deployments/PSDeploymentWhatIfCmdletParameters.cs:108]: Can you add a test for this?

Comment (Xynoclafe) [reply src/Resources/ResourceManager/SdkModels/Deployments/PSDeploymentWhatIfCmdletParameters.cs:108]: Pushed a test case

**Review by shenglol — APPROVED** [review]
:shipit:

**Review by shenglol — DISMISSED** [review]
Looks good to me but still need reviews from the PS team.

**Comment by Xynoclafe** [comment]
> Looks good to me but still need reviews from the PS team.

There are some credscan failures that I am trying to fix

**Comment by Xynoclafe** [comment]
> > Looks good to me but still need reviews from the PS team.
> 
> There are some credscan failures that I am trying to fix

The credscan issues are false positives as the credentials are invalid and are for non existent resources.

**Comment by Xynoclafe** [comment]
The credscan issues have been suppressed.
Unsure of the reason for test failures. The test fails claiming "Remove-AzStorageAccount" is not recognized as a cmdlet, which is weird. It is a valid cmdlet (https://docs.microsoft.com/en-us/powershell/module/az.storage/Remove-azStorageAccount?view=azps-5.5.0) and I was able to record the test successfully as well. So I don't know why a successful test record is failing during the checks. I would appreciate it if the reviewers could help if they've seen behavior like this before.

File: src/Resources/Resources.Test/ScenarioTests/ResourceTestRunner.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ protected ResourceTestRunner(ITestOutputHelper output)
                 .WithExtraRmModules(helper => new[]
                 {
                     helper.RMResourceModule,
-                    helper.GetRMModulePath("AzureRM.Monitor.psd1")
+                    helper.GetRMModulePath("AzureRM.Monitor.psd1"),
```
Comment (wyunchi-ms) [inline src/Resources/Resources.Test/ScenarioTests/ResourceTestRunner.cs:46]: Add `helper.GetRMModulePath("AzureRM.Storage.psd1"),` here to use cmdlets in Storage when running test cases.

Comment (Xynoclafe) [reply src/Resources/Resources.Test/ScenarioTests/ResourceTestRunner.cs:46]: @wyunchi-ms I made the suggested changes, but now it fails with "Device not configured"

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 [hunk 3]
```diff
@@ -884,4 +904,62 @@ function Test-TestDeploymentFromBicepFile
...
+
+		#Upload files to the StorageAccount
+		Set-AzStorageFileContent -ShareName "querystringshare" -Source "sampleLinkedTemplateParent.json" -Path "sampleLinkedTemplateParent.json" -Context $context
+		Set-AzStorageFileContent -ShareName "querystringshare" -Source "sampleLinkedTemplateChild.json" -Path "sampleLinkedTemplateChild.json" -Context $context
+
+		#Get SAStoken
+		$token = New-AzStorageAccountSASToken -Service File -ResourceType Service,Container,Object -Permission "r" -Context $context -ExpiryTime (Get-Date).AddMinutes(3)
+
+		#Create deployment
+		$deployment =New-AzResourceGroupDeployment -Name $rname -ResourceGroupName $rgname -TemplateUri "https://querystringpstests.file.core.windows.net/querystringshare/sampleLinkedTemplateParent.json" -QueryString $token.Substring(1)
```
Comment (wyunchi-ms) [inline src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:947]: Is this URL valid? I cannot access it in my environment.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 [hunk 4]
```diff
@@ -884,4 +904,62 @@ function Test-TestDeploymentFromBicepFile
...
+
+		# Assert
+		Assert-AreEqual Succeeded $deployment.ProvisioningState
+
+		#Run What-if
+		$result =  New-AzResourceGroupDeployment -Name $rname -ResourceGroupName $rgname -TemplateUri "https://querystringpstests.file.core.windows.net/querystringshare/sampleLinkedTemplateParent.json" -QueryString $token.Substring(1) -WhatIf
```
Comment (wyunchi-ms) [inline src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:953]: Is this URL valid? I cannot access it in my environment.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 [hunk 5]
```diff
@@ -797,17 +797,36 @@ function Test-NewDeploymentWithQueryString
...
+		#Upload files to the StorageAccount
+		Set-AzStorageFileContent -ShareName "querystringshare" -Source "sampleLinkedTemplateParent.json" -Path "sampleLinkedTemplateParent.json" -Context $context
+		Set-AzStorageFileContent -ShareName "querystringshare" -Source "sampleLinkedTemplateChild.json" -Path "sampleLinkedTemplateChild.json" -Context $context
+
+		#Get SAStoken
+		$token = New-AzStorageAccountSASToken -Service File -ResourceType Service,Container,Object -Permission "r" -Context $context -ExpiryTime (Get-Date).AddMinutes(2)
+
 		#Create deployment
-		$deployment =New-AzResourceGroupDeployment -Name $rname -ResourceGroupName $rgname -TemplateUri "https://querystringtesting.blob.core.windows.net/testqsblob/linkedTemplateParent.json" -QueryString "foo"
+		$deployment =New-AzResourceGroupDeployment -Name $rname -ResourceGroupName $rgname -TemplateUri "https://querystringpstests.file.core.windows.net/querystringshare/sampleLinkedTemplateParent.json" -QueryString $token.Substring(1)
```
Comment (wyunchi-ms) [inline src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:829]: Is this URL valid? I cannot access it in my environment.

Comment (Xynoclafe) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:829]: They are urls for a storage account created during the test and deleted after it is completed. So you will not be able to access it outside of the test. But at the time of the test execution (recording), the urls are valid.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 (hunk 4 above)
Comment (Xynoclafe) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:953]: Same reasoning as the comment below

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 (hunk 3 above)
Comment (Xynoclafe) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:947]: Same reasoning as the comment below

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 (hunk 5 above)
Comment (Xynoclafe) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1:829]: @wyunchi-ms any idea on what is happening with this error?

**Comment by Xynoclafe** [comment]
The checks should be passing now. @wyunchi-ms

**Comment by Xynoclafe** [comment]
The checks are either not starting or just getting stuck in running (even though the actual pipelines for them seem to have run successfully). Is there something wrong with the checks/pipeline?
Also, I see that credscan is failing. It seems like it is coming from  this file - https://github.com/Xynoclafe/azure-powershell/blob/xynoclafe/whatIfQueryString/src/CloudService/test/New-AzCloudService.Recording.json, based on https://dev.azure.com/azure-sdk/internal/_build/results?buildId=780021&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=a4822bc1-f3a9-5984-c941-6c1b7aa2ab81&l=18. 
This was last modified in this PR - https://github.com/Azure/azure-powershell/pull/14360 and was not modified in the current PR.
cc- @wyunchi-ms

**Comment by Xynoclafe** [comment]
The earlier check failures were due to some merge issues with the credScanSuppression file. The checks are passing now, we should be able to merge this.

**Review by wyunchi-ms — APPROVED** [review]

**Pull request merged**